typescript/xstate/src/promise.ts

```diff
@@ -0,0 +1,227 @@
+  PromiseActorLogic<TOutput, unknown>
+>;
+
+export function fromPromise<TOutput, TInput extends NonReducibleUnknown>(
+  promiseCreator: PromiseCreator<TOutput, TInput>,
+): PromiseActorLogic<TOutput, TInput> {
+  const logic: PromiseActorLogic<TOutput, TInput> = {
+    sentinel: "restate.promise.actor",
+    config: promiseCreator,
+    transition: (state, event) => {
+      if (state.status !== 'active') {
+        return state;
+      }
+
+      switch (event.type) {
+        case RESTATE_PROMISE_SENT: {
+          return {
+            ...state,
+            sent: true,
+          };
+        }
+        case RESTATE_PROMISE_RESOLVE: {
+          const resolvedValue = (event as any).data;
+          return {
+            ...state,
+            status: 'done',
+            output: resolvedValue,
+            input: undefined
+          };
+        }
+        case RESTATE_PROMISE_REJECT:
+          return {
+            ...state,
+            status: 'error',
+            error: (event as any).data,
+            input: undefined
+          };
+        case XSTATE_STOP:
+          return {
+            ...state,
+            status: 'stopped',
+            input: undefined
+          };
+        default:
+          return state;
+      }
+    },
+    start: (state, {self, system}) => {
+      if (state.status !== 'active') {
+        return;
+      }
+
+      if (state.sent) {
+        return;
+      }
+
+      const rs = (system as RestateActorSystem<any>)
+
+      rs.ctx.send(rs.api.promise).invoke({
+        systemName: rs.systemName,
+        self: serialiseActorRef(self),
+        srcs: actorSrc(self),
+        input: state.input
+      });
+
+      // note that we sent off the promise so we don't do it again
+      (system as any)._relay(self, self, {
+        type: RESTATE_PROMISE_SENT,
+      });
+    },
+    getInitialSnapshot: (_, input) => {
+      return {
+        status: 'active',
+        output: undefined,
+        error: undefined,
+        input,
+        sent: false,
+      };
+    },
+    getPersistedSnapshot: (snapshot) => snapshot,
+    restoreSnapshot: (snapshot: any) => snapshot
+  };
+
+  return logic;
+}
+
+function actorSrc(actor?: AnyActorRef): string[] {
+  if (actor === undefined) {
+    return []
+  }
+  if (typeof actor.src !== "string") {
+    return []
+  }
+  return [actor.src, ...actorSrc(actor._parent)]
+}
+
+export const promiseMethods = <TLogic extends AnyStateMachine>(path: string, logic: TLogic) => {
+  const api = xStateApi(path)
+
+  return {
+    invoke: async (ctx: restate.RpcContext, {
+      systemName,
+      self,
+      srcs,
+      input
+    }: { systemName: string, self: SerialisableActorRef, srcs: string[], input: NonReducibleUnknown }) => {
+      console.log("run promise with srcs", srcs, "in system", systemName, "with input", input)
+
+      const [promiseSrc, ...machineSrcs] = srcs
+
+      let stateMachine: AnyStateMachine = logic;
+      for (const src of machineSrcs) {
```

Comment:
apply the 'path' through the state machine definitions, checking for each one in either the implementations or directly in the defintiion